Cannot use allure.dynamic labels together with pytest.mark.parametrize

[//]: # (
. Note: for support questions, please use Stackoverflow or Gitter**. 
. This repository's issues are reserved for feature requests and bug reports.
.
. In case of any problems with Allure Jenkins plugin** please use the following repository 
. to create an issue: https://github.com/jenkinsci/allure-plugin/issues
.
. Make sure you have a clear name for your issue. The name should start with a capital 
. letter and no dot is required in the end of the sentence. An example of good issue names:
.
. - The report is broken in IE11
. - Add an ability to disable default plugins
. - Support emoji in test descriptions
)

#### I'm submitting a ... 
  - [x] bug report
  - [ ] feature request
  - [ ] support request => Please do not submit support request here, see note at the top of this template.

#### What is the current behavior?
` AttributeError: 'TestResultContainer' object has no attribute 'labels'` when using for instance `allure.dynamic.feature()` in test which is parametrized with `@pytest.mark.parametrize` decorator.

#### If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem

Minimal example to reproduce the issue:

```python
# test_one.py
import allure
import pytest

@pytest.mark.parametrize('param', ('one', 'two'))
def test_first(param):
    allure.dynamic.feature('feature1')
```

#### What is the expected behavior?
Should be possible to use `allure.dynamic` labels together with `pytest.mark.parametrize`.

#### What is the motivation / use case for changing the behavior?


#### Please tell us about your environment:

- Allure version:        2.4.1
- Test framework:        pytest@3.2.3
- Allure adaptor:        allure-pytest@2.2.3b1

#### Other information 
```$ pip freeze
allure-pytest==2.2.3b1
allure-python-commons==2.2.3b1
attrs==17.2.0
enum34==1.1.6
pluggy==0.5.2
py==1.4.34
pytest==3.2.3
six==1.11.0
```

```
pytest --alluredir=allure-results
================================================================================ test session starts ================================================================================
platform linux -- Python 3.6.3, pytest-3.2.3, py-1.4.34, pluggy-0.4.0
rootdir: /home/jb/tmp/allure2, inifile:
plugins: allure-pytest-2.2.3b1
collected 2 items                                                                                                                                                                    

test_one.py FF

===================================================================================== FAILURES ======================================================================================
__________________________________________________________________________________ test_first[one] __________________________________________________________________________________

param = 'one'

    @pytest.mark.parametrize('param', ('one', 'two'))
    def test_first(param):
>       allure.dynamic.feature('feature1')

test_one.py:8: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
venv/lib/python3.6/site-packages/allure_commons/_allure.py:66: in feature
    Dynamic.label(LabelType.FEATURE, *features)
venv/lib/python3.6/site-packages/allure_commons/_allure.py:58: in label
    plugin_manager.hook.add_label(label_type=label_type, labels=labels)
venv/lib/python3.6/site-packages/pluggy/__init__.py:680: in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
venv/lib/python3.6/site-packages/pluggy/__init__.py:240: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
venv/lib/python3.6/site-packages/pluggy/__init__.py:234: in <lambda>
    methods, kwargs, specopts=hook.spec_opts, hook=hook
venv/lib/python3.6/site-packages/allure_pytest/listener.py:188: in add_label
    self.allure_logger.update_test(None, labels=Label(label_type, label))
venv/lib/python3.6/site-packages/allure_commons/reporter.py:63: in update_test
    self._update_item(uuid, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <allure_commons.reporter.AllureReporter object at 0x7f5b9cdf7a90>, uuid = None, kwargs = {'labels': Label(name=<LabelType.FEATURE: 'feature'>, value='feature1')}
item = TestResultContainer(uuid='fcfb95a6-20ae-4d41-b698-5ad2279d76c1', name=None, children=[], description=None, description...ents=[], parameters=[], start=1509109701603, stop=1509109701603)], afters=[], links=[], start=1509109701603, stop=None)
name = 'labels', value = Label(name=<LabelType.FEATURE: 'feature'>, value='feature1')

    def _update_item(self, uuid, **kwargs):
        item = self._items[uuid] if uuid else self._items[next(reversed(self._items))]
        for name, value in kwargs.items():
>           attr = getattr(item, name)
E           AttributeError: 'TestResultContainer' object has no attribute 'labels'

venv/lib/python3.6/site-packages/allure_commons/reporter.py:17: AttributeError
__________________________________________________________________________________ test_first[two] __________________________________________________________________________________

param = 'two'

    @pytest.mark.parametrize('param', ('one', 'two'))
    def test_first(param):
>       allure.dynamic.feature('feature1')

test_one.py:8: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
venv/lib/python3.6/site-packages/allure_commons/_allure.py:66: in feature
    Dynamic.label(LabelType.FEATURE, *features)
venv/lib/python3.6/site-packages/allure_commons/_allure.py:58: in label
    plugin_manager.hook.add_label(label_type=label_type, labels=labels)
venv/lib/python3.6/site-packages/pluggy/__init__.py:680: in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
venv/lib/python3.6/site-packages/pluggy/__init__.py:240: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
venv/lib/python3.6/site-packages/pluggy/__init__.py:234: in <lambda>
    methods, kwargs, specopts=hook.spec_opts, hook=hook
venv/lib/python3.6/site-packages/allure_pytest/listener.py:188: in add_label
    self.allure_logger.update_test(None, labels=Label(label_type, label))
venv/lib/python3.6/site-packages/allure_commons/reporter.py:63: in update_test
    self._update_item(uuid, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <allure_commons.reporter.AllureReporter object at 0x7f5b9cdf7a90>, uuid = None, kwargs = {'labels': Label(name=<LabelType.FEATURE: 'feature'>, value='feature1')}
item = TestResultContainer(uuid='bc5f557b-5f4d-4b15-b963-3f427056db47', name=None, children=[], description=None, description...ents=[], parameters=[], start=1509109701680, stop=1509109701681)], afters=[], links=[], start=1509109701680, stop=None)
name = 'labels', value = Label(name=<LabelType.FEATURE: 'feature'>, value='feature1')

    def _update_item(self, uuid, **kwargs):
        item = self._items[uuid] if uuid else self._items[next(reversed(self._items))]
        for name, value in kwargs.items():
>           attr = getattr(item, name)
E           AttributeError: 'TestResultContainer' object has no attribute 'labels'

venv/lib/python3.6/site-packages/allure_commons/reporter.py:17: AttributeError
============================================================================= 2 failed in 0.15 seconds ==============================================================================
```

--- a/allure-behave/src/listener.py
+++ b/allure-behave/src/listener.py
@@ -105,7 +105,10 @@ class AllureListener(object):
         self.flush_steps()
         status = scenario_status(scenario)
         status_details = scenario_status_details(scenario)
-        self.logger.update_test(uuid, stop=now(), status=status, statusDetails=status_details)
+        test_result = self.logger.get_test(uuid)
+        test_result.stop = now()
+        test_result.status = status
+        test_result.statusDetails = status_details
         self.logger.close_test(uuid)
         self.current_step_uuid = None
 

--- a/allure-pytest/src/listener.py
+++ b/allure-pytest/src/listener.py
@@ -73,7 +73,9 @@ class AllureListener(object):
         yield
 
         for name, value in item.callspec.params.items() if hasattr(item, 'callspec') else ():
-            self.allure_logger.update_test(uuid, parameters=Parameter(name, represent(value)))
+            test_result = self.allure_logger.get_test(uuid)
+            if test_result:
+                test_result.parameters.append(Parameter(name, represent(value)))
 
         test_case.labels.extend([Label(name=name, value=value) for name, value in allure_labels(item)])
         test_case.labels.extend([Label(name=LabelType.TAG, value=value) for value in pytest_markers(item)])
@@ -94,11 +96,14 @@ class AllureListener(object):
     @pytest.hookimpl(hookwrapper=True)
     def pytest_runtest_call(self, item):
         uuid = self._cache.get(item.nodeid)
-        self.allure_logger.update_test(uuid, start=now())
+        test_result = self.allure_logger.get_test(uuid)
+        if test_result:
+            test_result.start = now()
 
         yield
 
-        self.allure_logger.update_test(uuid, stop=now())
+        if test_result:
+            test_result.stop = now()
 
     @pytest.hookimpl(hookwrapper=True)
     def pytest_fixture_setup(self, fixturedef, request):
@@ -170,10 +175,13 @@ class AllureListener(object):
             if report.failed and status == Status.PASSED:
                 status = Status.BROKEN
 
-        if status_details:
-            self.allure_logger.update_test(uuid, status=status, statusDetails=status_details)
-        else:
-            self.allure_logger.update_test(uuid, status=status)
+        test_result = self.allure_logger.get_test(uuid)
+        if test_result:
+            if status_details:
+                test_result.status = status
+                test_result.statusDetails = status_details
+            else:
+                test_result.status = status
 
     @allure_commons.hookimpl
     def attach_data(self, body, name, attachment_type, extension):
@@ -185,12 +193,15 @@ class AllureListener(object):
 
     @allure_commons.hookimpl
     def add_link(self, url, link_type, name):
-        self.allure_logger.update_test(None, links=Link(link_type, url, name))
+        test_result = self.allure_logger.get_test(None)
+        if test_result:
+            test_result.links.append(Link(link_type, url, name))
 
     @allure_commons.hookimpl
     def add_label(self, label_type, labels):
-        for label in labels:
-            self.allure_logger.update_test(None, labels=Label(label_type, label))
+        test_result = self.allure_logger.get_test(None)
+        for label in labels if test_result else ():
+            test_result.labels.append(Label(label_type, label))
 
 
 class ItemCache(object):

--- a/allure-pytest/test/labels/bdd/dynamic_bdd_labels_test.py
+++ b/allure-pytest/test/labels/bdd/dynamic_bdd_labels_test.py
@@ -1,0 +1,30 @@
+import allure
+import pytest
+
+
+@allure.feature('first feature')
+def test_dynamic_feature():
+    """
+    >>> allure_report = getfixture('allure_report')
+    >>> assert_that(allure_report,
+    ...             has_test_case('test_dynamic_feature',
+    ...                           has_feature('first feature'),
+    ...                           has_feature('second feature')
+    ...             )
+    ... )
+    """
+    allure.dynamic.feature('second feature')
+
+
+@pytest.mark.parametrize('feature', ['first feature', 'second feature'])
+def test_parametrized_dynamic_feature(feature):
+    """
+    >>> allure_report = getfixture('allure_report')
+    >>> for feature in ['first feature', 'second feature']:
+    ...     assert_that(allure_report,
+    ...                 has_test_case('test_parametrized_dynamic_feature[{feature}]'.format(feature=feature),
+    ...                               has_feature(feature)
+    ...                 )
+    ...     )
+    """
+    allure.dynamic.feature(feature)

--- a/allure-pytest/test/links/dynamic_links_test.py
+++ b/allure-pytest/test/links/dynamic_links_test.py
@@ -1,3 +1,4 @@
+import pytest
 import allure
 
 LINK_1 = "https://github.com"
@@ -15,3 +16,17 @@ def test_dynamic_links():
     ...             ))
     """
     allure.dynamic.link(LINK_2)
+
+
+@pytest.mark.parametrize('link', (LINK_1, LINK_2))
+def test_parametrized_dynamic_links(link):
+    """
+    >>> allure_report = getfixture('allure_report')
+    >>> for link in [LINK_1, LINK_2]:
+    ...     assert_that(allure_report,
+    ...                 has_test_case('test_parametrized_dynamic_links[{link}]'.format(link=link),
+    ...                               has_link(link)
+    ...                 )
+    ...     )
+    """
+    allure.dynamic.link(link)

--- a/allure-pytest/tox.ini
+++ b/allure-pytest/tox.ini
@@ -44,7 +44,7 @@ commands =
 #    `tox -e demo -- -k test_single_feature_label` or
 #    `tox -e demo -- ./test/steps/`
 [testenv:demo]
-basepython = python3.5
+#basepython = python3.5
 
 passenv = HOME
 

--- a/allure-python-commons/src/reporter.py
+++ b/allure-python-commons/src/reporter.py
@@ -68,9 +68,6 @@ class AllureReporter(object):
     def get_test(self, uuid):
         return self.get_item(uuid) if uuid else self.get_last_item(TestResult)
 
-    #def update_test(self, uuid, **kwargs):
-    #    self._update_item(uuid, **kwargs)
-
     def close_test(self, uuid):
         test_case = self._items.pop(uuid)
         plugin_manager.hook.report_result(result=test_case)

--- a/allure-python-commons/src/reporter.py
+++ b/allure-python-commons/src/reporter.py
@@ -2,6 +2,7 @@ from collections import OrderedDict
 
 from allure_commons.types import AttachmentType
 from allure_commons.model2 import ExecutableItem
+from allure_commons.model2 import TestResult
 from allure_commons.model2 import Attachment, ATTACHMENT_PATTERN
 from allure_commons.utils import now
 from allure_commons._core import plugin_manager
@@ -27,6 +28,11 @@ class AllureReporter(object):
 
     def get_item(self, uuid):
         return self._items.get(uuid)
+
+    def get_last_item(self, item_type):
+        for _uuid in reversed(self._items):
+            if type(self._items[_uuid]) == item_type:
+                return self._items.get(_uuid)
 
     def start_group(self, uuid, group):
         self._items[uuid] = group
@@ -59,8 +65,11 @@ class AllureReporter(object):
     def schedule_test(self, uuid, test_case):
         self._items[uuid] = test_case
 
-    def update_test(self, uuid, **kwargs):
-        self._update_item(uuid, **kwargs)
+    def get_test(self, uuid):
+        return self.get_item(uuid) if uuid else self.get_last_item(TestResult)
+
+    #def update_test(self, uuid, **kwargs):
+    #    self._update_item(uuid, **kwargs)
 
     def close_test(self, uuid):
         test_case = self._items.pop(uuid)

